### ISSUE TITLE
__iterlen__ is broken on Python 3

### ISSUE THREAD

**Issue by polybassa** [secdev/scapy#3380]
### Brief description

I found this bug during my tests on #3379 

It seems, like `__iterlen__` is not returning the right value, if a generator is given to it. Here are some examples:

```
>>> pks = IP(version=[4,6])
>>> pks.__iterlen__()
2
>>> pks = IP(version=range(4,8,2))
>>> pks.__iterlen__()
1
>>> pks = IP(version=(4,6))
>>> pks.__iterlen__()
3
>>> pks = IP(version=(4,6,2))
>>> pks.__iterlen__()
1
>>> pks = IP(version=(x for x in [4,6]))
>>> pks.__iterlen__()
1

```

### Scapy version

current

### Python version

all versions of python 3

### Operating system

Linux, Windows, OSX

### How to reproduce

See #3379

**Comment by gpotter2** [comment]
Since when are generators supported in fields generation?

**Comment by polybassa** [comment]
Thanks for your answer. 
I expected generators to be supported in fields generation, since they work as expected in python2. 
```
>>> pks = IP(version=range(4,8,2))
>>> for p in pks:
...     print(repr(p))
... 
<IP  version=4 |>
<IP  version=6 |>
>>> pks.__iterlen__()
2
```
python3
```
>>> pks = IP(version=range(4,8,2))
...: for p in pks:
...:     print(repr(p))
...: 
<IP  version=4 |>
<IP  version=6 |>
>>> pks.__iterlen__()
1
```

**Comment by polybassa** [comment]
Having python2 behave different than python3 is a bug, in my opinion. 
If only some special cases are supported, I would see two possibilities for a fix.
Either raise an error, if a random generator is passed, which doesn't match one of the special cases, or let python2 and python3 behave identical.

**Comment by guedou** [comment]
Thanks for opening the discussion on this issue.

What are the benefits and drawbacks of using generators as field values? Also, could you describe how you discovered this behavior?

**Comment by polybassa** [comment]
Hi, thanks for your response.

Benefits: more flexibility for future use cases
Drawbacks: Generators have to be handled with car, since they can only drained once. This makes especially the __iterlen__ function complicated. If a generator is transformed into a list, performance or memory problems could occur.

How I discovered this.

Magically, generators and iterators are already working for packet generation, as long only the __iter__ function of Packet is used. Basically as long as you treat a Packet instance like a generator, everything is fine. 

I always thought this is an intended feature of scapy, because it worked so naturally. 

Therefore I used this in #3379 to create packets for my test.

Now I encountered strange errors on python3 when I passed such a packet object to sr().
sr () uses __iterlen__ to evaluate how many packets need to be received. notans is the variable for that.
On python2 everything worked as expected in my tests, since isinstance(val, list) inside __iterlen__ evaluates to True if val is a range. On python3 this changed, which finally broke my tests.

To summarize: extending __iterlen__ to support range is easy and should not cause drawbacks.
Generators as field value are partly supported. In order to prevent other users from unintended behavior, we could forbid generators, convert them to lists, or carefully investigate all possible drawbacks. Basically a generator as field value would convert a Packet to a PacketGenerator and then we could check if all current uses of Packet are also able to deal with PacketGenerator objects. 




-- 
Viele Grüße

Nils Weiß

**Comment by guedou** [comment]
Thanks for this detailed answer!

**Comment by gpotter2** [comment]
Thanks for the analysis. I'm fine with allowing ranges as long as we explicitly forbidding generators.

**Comment by polybassa** [comment]
Another question, what about removing __iterlen__ entirely? 
AFAIK, it's only used inside Packet.__iter__ and sr(). 

-- 
Viele Grüße

Nils Weiß

**Comment by gpotter2** [comment]
You can try to come up with an alternative but it's pretty efficient. We need to know how many packets are generated from one packet pretty early on in `sr()` and turning a packet generator into a giant list is not an option.

**Comment by polybassa** [comment]
I totally agree on that. I thought about some way to share the iterator finished information with the receiver thread, but I did not fully checked the code, yet. I will have a look and check if it is possible to remove. 
-- 
Viele Grüße

Nils Weiß

**Comment by polybassa** [comment]
Another question, just to not provide a wrong solution or to make a false assumption. Is there a use case of sndrecv where the receiver thread has to be stopped before the sender has sent the last packet? 
-- 
Viele Grüße

Nils Weiß

**Comment by guedou** [comment]
I think that could be useful for sr*loop() functions
